docker/archive/dest.go

```diff
@@ -70,3 +75,61 @@ func (d *archiveImageDestination) Close() error {
 func (d *archiveImageDestination) Commit(ctx context.Context, unparsedToplevel types.UnparsedImage) error {
 	return d.Destination.Commit(ctx)
 }
+
+func (m multiImageDestinationReference) NewImageDestination(_ context.Context, _ *types.SystemContext) (types.ImageDestination, error) {
+	return m.dest, nil
+}
+
+// MultiImageDestinations allows for creating and writing to docker archives
+// that include more than one image.
+type MultiImageDestination struct {
+	*archiveImageDestination
+	path string
+}
+
+// multiImageDestinationReference is a types.ImageReference embedding a MultiImageDestination.
+type multiImageDestinationReference struct {
+	*archiveReference
+	dest *MultiImageDestination
+}
+
+// NewMultiImageDestination returns a MultiImageDestination for the specified path.
+func NewMultiImageDestination(sys *types.SystemContext, path string) (*MultiImageDestination, error) {
+	dest, err := newArchiveImageDestination(sys, path, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &MultiImageDestination{dest, path}, nil
+}
+
+// Reference returns an ImageReference embedding the MultiImageDestination.
+func (m *MultiImageDestination) Reference() types.ImageReference {
+	ref := &archiveReference{path: m.path}
```

Comment:
This does not include the tag

Reply:
Tags are set via `AdditionalTags`.

https://github.com/containers/libpod/pull/6811/files#diff-3303ff4a5ad91328550c7bee8df0be69R259

Reply:
Yes, but not on the reference = not in `copy.Image` error messages.

Reply:
OTOH there is still the case of saving untagged images, so references don’t _always_ have any extra data anyway.

I’ll read all of this more carefully a bit later.

Reply:
Much appreciated, thanks!

I am on PTO tomorrow but would will go back to this on Monday morning. Ideally, we need to get the feature in next week.

Reply:
BTW `AdditionalTags` was basically a hack to allow creating archives with multiple tags from a single image as a workaround for not having true multi-image support (where the same result would be obtained naturally by just copying the image to references with different tags); so this PR should conceptually eliminate the need for it.

Reply:
@mtrmac, ideas how to proceed?

Reply:
@mtrmac @rhatdan ... can we get this moving or are we blocked on something? I am getting increasing pressure to get this done.

Reply:
Can we please proceed?

Reply:
I’m sorry for not getting back, I did promise I would.

Still, https://github.com/containers/image/pull/975#pullrequestreview-441846261 was I think fairly clear to the general direction. See https://github.com/containers/image/pull/991 for an unfinished prototype of the read side. Yes, the PR is larger, but it already includes the ability to read any (even untagged) image in an archive using a textual reference, and a lot of the “new” code is actually only moved — `tarfile.Source` was just split into two, the only non-trivial _net new_ code at that layer is just `chooseManifest`.

Reply:
Thanks! I am still not sure how to proceed.

Now we have two PRs and I am worried we are running out of time; in fact, we might not get it into RHEL 8.3 any more.

If you agree, I want to focus on Podman-only needs first. We can still make follow-up cards for a more generally applicable solution, in case that will buy us some time. Certainly, the API shouldn't break.